File: 2014-04-18-FP/SubsetSum/ss.accepted.fs
open System

type T = L | N of (T * int64 * int64 * T)

let rec mktree (n : int64) xs =
        match (n, xs) with
        | (c, xs) when c = 0L -> (L, xs)
        | (c, (n, s)::xs) when c = 1L -> (N (L, n, s, L), xs)
        | (n, xs) ->
                let (midp : int64) = (n + 1L) / 2L
                let (l, ((n1, s1)::xs1)) = mktree (midp - 1L) xs
                let (r, xs2) = mktree (n - midp) xs1
                (N (l, n1, s1, r), xs2)

let rec tfind t x : int64 =
        match (t, x) with
        | (L, _) -> -1L
        | ((N (l, n, s, r)), x) ->
                if x = s
                    then n
                    else
                        if x < s
                            then
                                let l1 = tfind l x
                                if l1 = -1L
                                    then n
                                    else l1
                            else tfind r x

let tst t =
        let s = int64 (Console.ReadLine ())
        Console.WriteLine ("{0}", tfind t s)

let main _ =
        ignore (Console.ReadLine ())
        let str = Console.ReadLine ()
        let xs = List.map (fun x -> int64 x) (List.ofSeq (str.Split [|' '|]))
        let ys = List.rev (List.sort xs)
        let (_, y0s) = List.fold (fun (s, l) x -> (s + x, (s + x)::l)) (0L, []) ys
        let y1s = List.zip [1L .. (int64 (List.length xs))] (List.rev y0s)
        let t = int64 (Console.ReadLine ())
        let (tree, _) = mktree (int64 (List.length xs)) y1s
        ignore (List.map (fun _ -> tst tree) [1L .. t])

main ()

File: 2014-04-18-FP/SubsetSum/ss.accepted.scala
object Solution {
    sealed abstract class STree
    case object Leaf extends STree
    case class Node(l: STree, n: Long, s: Long, r: STree) extends STree

    def mktree(n: Long, xs: List[(Long, Long)]): (STree, List[(Long, Long)]) = (n, xs) match {
        case (0, xs) => (Leaf, xs)
        case (1, (n, s)::xs) => (Node(Leaf, n, s, Leaf), xs)
        case (n, xs) => {
            val midp = (n + 1) / 2
            val (l, ((n1, s1)::xs1)) = mktree(midp - 1, xs)
            val (r, xs2) = mktree(n - midp, xs1)
            (Node(l, n1, s1, r), xs2)
        }
    }

    def tfind(t: STree, x: Long): Long = (t, x) match {
        case (Leaf, _) => -1
        case (Node(l, n, s, r), x) => 
            if (x == s)
                n
            else if (x < s) {
                val l1 = tfind(l, x)
                if (l1 == -1)
                    n
                else
                    l1
            }
            else
                tfind(r, x)
    }

    def tst(t: STree): Unit = {
        val s = readLine.toLong
        println(tfind(t, s))
    }

    def main(args: Array[String]) = {
        readLine
        val xs = readLine.split(" ").map(_.toLong).sortWith(_ > _)
        def f(z: (Long, List[Long]), x: Long): (Long, List[Long]) =
            (z._1 + x, (z._1 + x)::(z._2))
        val (_, ys) = (xs foldLeft (0L, List[Long]()))(f)
        val y0s = (1L to ys.length.toLong).toList zip (ys.reverse)
        val (tree, _) = mktree(y0s.length.toLong, y0s)
        val t = readLine.toLong
        for (ix <- 0L until t) {
            tst(tree)
        }
    }
}
